File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    var answer = 0;
    words.push(begin)
    const graph = Array(words.length).fill().map(()=>[])
    

    for (let i=0; i<words.length-1; i++){
        for (let j=i+1; j<words.length; j++){
            if (isConnected(words[i],words[j])){
                graph[i].push(j)
                graph[j].push(i)
            }
        }
    }
    
    const visited = Array(words.length).fill(false)

    let answers = []
    
    function dfs(node, target, count){
        visited[node] = true
        if (node === target) answers.push(count)
        for (let next of graph[node]){
            if (!visited[next]) {
                dfs(next,target,count+1)
                visited[next] = false
            }
        }
        
    }

    let start = words.indexOf(begin)
    let end = words.indexOf(target)
    dfs(start, end, 0)
    return answers.length ? Math.min(...answers) : 0;
}

const isConnected = (w1, w2) => {
    let diff = 0
    for (let i=0; i<w1.length; i++){
        if (w1[i] !== w2[i]) diff++
    }
    return diff === 1
}

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    visited = set()
    global answer
    answer = float('inf')
    def dfs(curr,count):
        global answer
        if curr == target:
            answer = min(answer, count)
            return
        for w in words:
            if is_connect(curr, w) and w not in visited:
                visited.add(w)
                dfs(w,count+1)
                visited.pop()
                
    def is_connect(a,b):
        count = 0
        for i in range(len(a)):
            if a[i] != b[i]:
                count += 1
        return count == 1

    if target not in words:
        return 0
    dfs(begin,0)
    return answer 
                
    
